File: Python/Problem_091.py
import math

def max_steps(x, y, dx, dy, N):
    INF = 10**18
    lim = INF
    if dx > 0: lim = min(lim, (N - x) // dx)
    elif dx < 0: lim = min(lim, x // (-dx))
    if dy > 0: lim = min(lim, (N - y) // dy)
    elif dy < 0: lim = min(lim, y // (-dy))
    return 0 if lim == INF else lim

def euler_91(N=50):
    ans = N * N  # right angle at origin
    for x in range(N + 1):
        for y in range(N + 1):
            if x == 0 and y == 0:
                continue
            g = math.gcd(x, y)
            dx = -y // g
            dy =  x // g
            ans += max_steps(x, y, dx, dy, N)
            ans += max_steps(x, y, -dx, -dy, N)
    return ans

if __name__ == "__main__":
    print(euler_91(50))  # 14234

File: Java/Problem_091.java
public class Euler91 {
    static long maxSteps(int x, int y, int dx, int dy, int N) {
        long INF = (1L << 60);
        long lim = INF;

        if (dx > 0) lim = Math.min(lim, (long)(N - x) / dx);
        else if (dx < 0) lim = Math.min(lim, (long)x / (-dx));

        if (dy > 0) lim = Math.min(lim, (long)(N - y) / dy);
        else if (dy < 0) lim = Math.min(lim, (long)y / (-dy));

        return (lim == INF) ? 0 : lim;
    }

    static int gcd(int a, int b) {
        while (b != 0) {
            int t = a % b;
            a = b;
            b = t;
        }
        return a;
    }

    public static void main(String[] args) {
        final int N = 50;
        long ans = 1L * N * N; // right angle at origin

        for (int x = 0; x <= N; x++) {
            for (int y = 0; y <= N; y++) {
                if (x == 0 && y == 0) continue;

                int g = gcd(x, y);
                int dx = -y / g;
                int dy =  x / g;

                ans += maxSteps(x, y, dx, dy, N);
                ans += maxSteps(x, y, -dx, -dy, N);
            }
        }

        System.out.println(ans); // 14234
    }
}
